src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.Split.cs

```
@@ -0,0 +1,136 @@
+
+            public SpanSplitEnumerator<T> GetEnumerator() => this;
+
+            public Range Current => new Range(_startCurrent, _endCurrent);
+
+            internal SpanSplitEnumerator(ReadOnlySpan<T> span, SearchValues<T> searchValues)
+            {
+                _span = span;
+                _separatorLength = 1;
+                _splitMode = SplitMode.SearchValues;
+                _searchValues = searchValues;
+            }
+
+            internal SpanSplitEnumerator(ReadOnlySpan<T> span, ReadOnlySpan<T> separator, bool treatAsSingleSeparator)
+            {
+                _span = span;
+                _separatorBuffer = separator;
+                _separatorLength = (_separatorBuffer.Length, treatAsSingleSeparator) switch
+                {
+                    (0, true) or (_, false) => 1,
+                    _ => separator.Length
+                };
+                _splitMode = treatAsSingleSeparator ? SplitMode.Sequence : SplitMode.Any;
+            }
+
+            internal SpanSplitEnumerator(ReadOnlySpan<T> span, T separator)
+            {
+                _span = span;
+                _separator = separator;
+                _separatorLength = 1;
+                _splitMode = SplitMode.SingleToken;
+            }
+
+            public bool MoveNext()
+            {
+                if (!_isInitialized || _startNext > _span.Length)
+                {
+                    return false;
+                }
+
+                ReadOnlySpan<T> slice = _span[_startNext..];
+
+                int separatorIndex = _splitMode switch
+                {
+                    SplitMode.SingleToken => slice.IndexOf(_separator),
+                    SplitMode.Sequence => slice.IndexOf(_separatorBuffer),
+                    SplitMode.Any => slice.IndexOfAny(_separatorBuffer),
+                    SplitMode.SearchValues => _searchValues.IndexOfAny(_span),
+                    _ => throw new UnreachableException()
```

Should this throw or just silently fail?

---

How would you get here such that it would matter? Assuming it's truly unreachable, I'd instead have an assert on the split mode before the switch and then just use _ for one of the valid cases.